abci/app.go

```diff
@@ -0,0 +1,377 @@
+	return &types.ResponseCheckTx{Code: CodeTypeOK}, nil
+}
+
+// Consensus Connection
+// Initialize blockchain w validators/other info from CometBFT
+func (app ForumApp) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	for _, v := range req.Validators {
+		app.updateValidator(v)
+	}
+	appHash := app.state.Hash()
+
+	// This parameter can also be set in the genesis file
+	req.ConsensusParams.Abci.VoteExtensionsEnableHeight = 1
+	return &abci.ResponseInitChain{ConsensusParams: req.ConsensusParams, AppHash: appHash}, nil
+}
+
+func (app *ForumApp) PrepareProposal(_ context.Context, proposal *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	fmt.Println("entered prepareProp")
+
+	voteExtensionCurseWords := app.getWordsFromVe(proposal.LocalLastCommit.Votes)
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			proposedTxs = append(proposedTxs, tx)
+		}
+		// Adding the curse words from vote extensions too
+		if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+			proposedTxs = append(proposedTxs, tx)
+		} else {
+			banTx := model.BanTx{UserName: msg.Sender}
+			bannedUsersString[msg.Sender] = struct{}{}
+			resultBytes, err := json.Marshal(banTx)
+			if err == nil {
+				finalProposal = append(finalProposal, resultBytes)
+			} else {
+				panic(fmt.Errorf("invalid ban transaction in prepareProposal"))
+			}
+		}
+	}
+
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			panic(err)
+		}
+		if _, ok := bannedUsersString[msg.Sender]; !ok {
+			finalProposal = append(finalProposal, tx)
+		}
+	}
+	return &types.ResponsePrepareProposal{Txs: finalProposal}, nil
+}
+
+func (ForumApp) ProcessProposal(_ context.Context, processproposal *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+	fmt.Println("entered processProp")
+	bannedUsers := make(map[string]struct{}, 0)
+
+	finishedBanTxIdx := len(processproposal.Txs)
+	for i, tx := range processproposal.Txs {
+		if isBanTx(tx) {
+			var parsedBan model.BanTx
+			err := json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for _, tx := range processproposal.Txs[finishedBanTxIdx:] {
+		msg, err := model.ParseMessage(tx)
```

Comment:
You first need to check `isBanTx` and it it is a ban Tx, you reject

Reply:
If it is a `BanTx` it will fail to parse using `ParseMessage` because I don't append a string but rather create a completely different transaction type.

Reply:
Please add a comment explaining it (it's not clear IMO)

Reply:
Done.